### PR TITLE
Feat/persistent nodes

### PR DESCRIPTION
**Pull request by bludnic**
image

- [x] Persistent node checkbox state after page refresh
- [x] Persistent `useFastest` after page refresh
- [x] Allow navigating to Nodes page from Login page

image

### CODE REVIEW

**Comment by bludnic** [comment]
[Feat: Save Nodes list in Browser's storage](https://trello.com/c/MrPCLp9j/221-feat-save-nodes-list-in-browsers-storage)

**Comment by bludnic** [comment]
Merge `dev` into the current branch after #547 merge.

File: src/lib/nodes/storage.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+function getOptions(): Options {
+  return getItem(NODES_OPTIONS_STORAGE_KEY, {
+    adm: { useFastest: false },
+    btc: { useFastest: false },
+    doge: { useFastest: false },
+    dash: { useFastest: false },
+    eth: { useFastest: false },
+    lsk: { useFastest: false }
```
[the remark this answers is not in the document]
Comment (bludnic) [reply src/lib/nodes/storage.ts]: It will lose in readability. Moved `defaultValue` to const.

**Comment by adamant-al** [comment]
- Don't shift other UI elements with the Settings icon in the Login screen, and place it under the language selector, both aligned right
- Add `Use socket connections` to storage also
- Move `useFastest` default values to `adamant-wallets`, false for `adm`, true for other nodes and services
- Don't request disabled nodes on app start
- When you open an app with Nodes screen, e.g., https://adamant-gcx03h8yu-adamant-team.vercel.app/options/nodes, the Back button doesn't work. Either redirect to the Login screen, or go to Login screen with the Back button.
- Either don't show the Socket cell for inactive nodes, or make the v/x mark gray.

**Comment by bludnic** [comment]
> Add Use socket connections to storage also

It is already stored in the storage.

> Move useFastest default values to adamant-wallets, false for adm, true for other nodes and services

Don't see a reason to do this. It seems like an app-specific setting. Even if we add it to adamant-wallets it will be fetched  once and saved to LocalStorage. Next app reloads will use cached value. We will not be able to control that value from adamant-wallets. It will also slow the development flow.

**Comment by adamant-al** [comment]
> Don't see a reason to do this. It seems like an app-specific setting. Even if we add it to adamant-wallets it will be fetched once and saved to LocalStorage. Next app reloads will use cached value. We will not be able to control that value from adamant-wallets. It will also slow the development flow.

For example, we have `defaultVisibility` for coin wallets, why not have default settings for node groups?

We have many ADM nodes and they are generally fast. Using a random node will leverage privacy.

Opposite for coin nodes, they are slow and we can balance their load with the default `Use the fastest node` setting.

**Comment by just-software-dev** [comment]
I think for this moment it's not necessary to include "use the fastest node" to the adamant-wallets. It's a very small setting and we can use `True` by default for all node groups. Currently we are using `True` by default on iOS

**Comment by adamant-al** [comment]
Ok, keep it as is.

**Comment by bludnic** [comment]
> Don't request disabled nodes on app start

Healtcheck is not trigerred for disabled nodes. You probably see requests to `GET /api/node/status` on app start. This is not a healtcheck. This endpoint is called only [once](https://github.com/Adamant-im/adamant-im/blob/dev/src/lib/nodes/adm/AdmNode.ts#L44), just to get the info about the node (version, socketSupport, wsPort)

**Comment by adamant-al** [comment]
> > Don't request disabled nodes on app start
> 
> Healtcheck is not trigerred for disabled nodes. You probably see requests to `GET /api/node/status` on app start. This is not a healtcheck. This endpoint is called only [once](https://github.com/Adamant-im/adamant-im/blob/dev/src/lib/nodes/adm/AdmNode.ts#L44), just to get the info about the node (version, socketSupport, wsPort)

It's important not triggering disabled nodes because of privacy concerns.
@bludnic @just-software-dev

**Comment by bludnic** [comment]
> It's important not triggering disabled nodes because of privacy concerns.

Skipped `fetchNodeInfo` for disabled nodes

**Comment by bludnic** [comment]
> When you open an app with Nodes screen, e.g., [https://adamant-gcx03h8yu-adamant-team.vercel.app/options/nodes](http://localhost:8080/options/nodes), the Back button doesn't work. Either redirect to the Login screen, or go to Login screen with the Back button.

Fixed. User will be redirected to the main page if there are no pages in `window.history`

**Comment by bludnic** [comment]
> Either don't show the Socket cell for inactive nodes, or make the v/x mark gray.

Fixed. Removed socket column. Also fixed status column.

**Comment by bludnic** [comment]
@adamant-al @martiliones check now

**Comment by adamant-al** [comment]
- Don't shift other UI elements with the Settings icon in the Login screen
- Reduce vertical padding by 8px

Screenshot 2023-11-14 в 16 32 34

Others are good.

**Comment by bludnic** [comment]
@adamant-al We can place the setting icon on the left side. This way nothing is shifted.

https://github.com/Adamant-im/adamant-im/assets/25831507/7dd771f8-0c2c-4565-a55a-95a414d0bf7d

**Comment by adamant-al** [comment]
No, put it as it now, to the right.
Set span height to 0 or approach another way not to shift.

**Comment by bludnic** [comment]
> No, put it as it now, to the right. Set span height to 0 or approach another way not to shift.

image
image

Must note that the icon button has hover/ripple effect. In reality there are no vertical space between these buttons.
I can decrease vertical space only by making the setting icon smaller.

Normal:
image

Hover:
image
image

**Comment by adamant-al** [comment]
- `<div data-v-a80aa572="" class="text-right">`:  height to `0` 
- `<button data-v-a80aa572="" type="button" class="v-btn `:  height from `calc(var(--v-btn-height) + 12px)` = 48 to `32px`

**Comment by bludnic** [comment]
@adamant-al fixed

**Comment by adamant-al** [comment]
Looks wrong.
Apply styles here https://adamant-fh5rc7p15-adamant-team.vercel.app/
- `<div data-v-a80aa572="" class="text-right">`:  height to `0` 
- `<button data-v-a80aa572="" type="button" class="v-btn `:  height from `calc(var(--v-btn-height) + 12px)` = 48 to `32px`

**Comment by bludnic** [comment]
@adamant-al This is an old build. Check the actual link commented by Vercel bot.
https://adamant-im-git-feat-persistent-nodes-adamant-team.vercel.app/

image

**Comment by adamant-al** [comment]
@bludnic 
Look at my commit; move `style="height: 0"` and `style="margin-right: 8px"` to CSS.

**Comment by bludnic** [comment]
> @bludnic Look at my commit; move `style="height: 0"` and `style="margin-right: 8px"` to CSS.

Rewrited you solution by using `position: absolute`. It looks the same.

**Comment by adamant-al** [comment]
@martiliones @just-software-dev @RealGoodProgrammer 
Review and merge.

**Pull request merged**